core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,312 @@
+    return aggregate
+        .getInput()
+        .accept(this)
+        .map(input -> ImmutableAggregate.builder().from(aggregate).input(input).build());
+  }
+
+  @Override
+  public Optional<Rel> visit(Fetch fetch) throws RuntimeException {
+    return fetch
+        .getInput()
+        .accept(this)
+        .map(input -> ImmutableFetch.builder().from(fetch).input(input).build());
+  }
+
+  @Override
+  public Optional<Rel> visit(Filter filter) throws RuntimeException {
+    var input = filter.getInput().accept(this);
+    var condition = visitExpression(filter.getCondition());
+    if (allEmpty(input, condition)) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        ImmutableFilter.builder()
+            .from(filter)
+            .input(input.orElse(filter.getInput()))
+            .condition(condition.orElse(filter.getCondition()))
+            .build());
+  }
+
+  @Override
+  public Optional<Rel> visit(Join join) throws RuntimeException {
+    var left = join.getLeft().accept(this);
+    var right = join.getRight().accept(this);
+    var condition = join.getCondition().flatMap(t -> visitExpression(t));
+    var postFilter = join.getPostJoinFilter().flatMap(t -> visitExpression(t));
+    if (allEmpty(left, right, condition, postFilter)) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        ImmutableJoin.builder()
+            .from(join)
+            .left(left.orElse(join.getLeft()))
+            .right(right.orElse(join.getRight()))
+            .condition(condition.or(() -> join.getCondition()))
+            .postJoinFilter(postFilter.or(() -> join.getPostJoinFilter()))
+            .build());
+  }
+
+  @Override
+  public Optional<Rel> visit(Project project) throws RuntimeException {
+    var input = project.getInput().accept(this);
+    Optional<List<Expression>> expressions = copyOnWrite(project.getExpressions());
+    if (allEmpty(input, expressions)) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        ImmutableProject.builder()
+            .from(project)
+            .input(input.orElse(project.getInput()))
+            .expressions(expressions.orElse(project.getExpressions()))
+            .build());
+  }
+
+  @Override
+  public Optional<Rel> visit(Sort sort) throws RuntimeException {
+    return sort.getInput()
+        .accept(this)
+        .map(input -> ImmutableSort.builder().from(sort).input(input).build());
+  }
+
+  public Optional<Rel> visitRel(Rel rel) {
+    return rel.accept(this);
+  }
+
+  private Optional<Expression> visitExpression(Expression expression) {
+    ExpressionVisitor<Optional<Expression>, RuntimeException> visitor =
+        new AbstractExpressionVisitor<>() {
+          @Override
+          public Optional<Expression> visitFallback(Expression expr) {
+            return Optional.empty();
+          }
+
+          @Override
+          public Optional<Expression> visit(Expression.Switch expr) throws RuntimeException {
```

Comment:
you can remove these throws RuntimeExceptions. This is always implied by Java already.

Reply:
I'll follow up with another PR for this. There are 84 occurrences across 9 files. This happens when the methods are generated by the IDE.